### ISSUE TITLE
Unsubscribe tags workaround NOT working on Transactional Emails (see #59)

### ISSUE THREAD

**Issue by rg34** [sendgrid/sendgrid-ruby#143]
Hello. 

I've been using the workaround as described in the issue below. However, it is no longer working. I can verify that it stopped working somewhere after January 23, 2017.
#59 

I have tried every combination of saving in code view vs design view with no luck. Any word on the ETA for an actual fix?

Thanks so much!

![screen shot 2017-02-03 at 8 34 09 am](https://cloud.githubusercontent.com/assets/4340339/22597900/c127770c-e9ee-11e6-8ae3-c57d263e9331.png)

**Comment by thinkingserious** [comment]
Hi @rg34,

What exactly is happening now when you save?

Have you tried using different tags? e.g. -asm_preferences_url-

**Comment by rg34** [comment]
codeview: 
![screen shot 2017-02-03 at 11 56 09 am](https://cloud.githubusercontent.com/assets/4340339/22604246/daa7742a-ea07-11e6-836a-d3706a6d0e84.png)

email view: 
![screen shot 2017-02-03 at 11 55 53 am](https://cloud.githubusercontent.com/assets/4340339/22604253/e2bbb964-ea07-11e6-952e-4ec7e0f68492.png)

I can verify in an email sent January 23, that the workaround was still functional: 
![screen shot 2017-02-03 at 8 48 27 am](https://cloud.githubusercontent.com/assets/4340339/22604271/000f0304-ea08-11e6-88a6-7af6a9565f14.png)

I will try sending the substitution tag, but won't get to that for a couple of hours. I'll let you know. 

Thanks!

**Comment by thinkingserious** [comment]
Hi @rg34,

Did changing the substitution tag help?

**Comment by soundornoise** [comment]
Hi @thinkingserious,

I think I am hitting this same problem. I'm creating a new template, so I can't comment specifically on the change in behaviour, but I am seeing the tags getting encoded.

I thought the asm tags were preset, were you suggesting having a substitution tag map to <%asm_preferences_url%>? I attempted this and, while the tags are not encoded, they do not get transformed in to the URL.

@rg34 Does this match what you've seen?

Thanks

**Comment by thinkingserious** [comment]
@soundornoise,

I'm digging deeper with our support team and I'll post the results of my findings here.

Thanks!

**Comment by kylearoberts** [comment]
@soundornoise @rg34 

I work in the support department and wanted to help get this situation cleared up. Can you confirm that you are just looking for the URL for the unsubscribe link? Are you looking to do something like:

`<a href="<%asm_preferences_url%>" >Preferences</a>`

Or are you having issues using substitutions:

```
  personalization = Personalization.new
  personalization.to = Email.new(email: 'test1@example.com', name: 'Example User')
  personalization.substitutions = Substitution.new(key: '%asm_group_unsubscribe_url%', value: '<%asm_group_unsubscribe_url%>')
  personalization.substitutions = Substitution.new(key: '%asm_global_unsubscribe_url%', value: '<%asm_global_unsubscribe_url%>')
  personalization.substitutions = Substitution.new(key: '%asm_preferences_url%', value: '<%asm_preferences_url%>')
  mail.personalizations = personalization
```

Let me know if I have not understood correctly. Once I know more I will better be able to direct you.

**Comment by soundornoise** [comment]
Hi @kylearoberts,

Thanks for following up. Yes, I'd like to add the preferences link as you described. I'm currently using a custom unsubscribe tag mapped to `[unsubscribe]` to avoid encoding issues, but adding the preferences link would be ideal.

What I have:
~~~~
  <a href="<%asm_preferences_url%>" >Preferences</a> |
  <a href="<%asm_group_unsubscribe_url%>" >Unsub</a> |
  <a href="<%asm_global_unsubscribe_url%>" >ALL</a>
~~~~

On save gets re-encoded to:

~~~~
<a href="&lt;%asm_preferences_url%&gt;">Preferences</a> | <a href="&lt;%asm_group_unsubscribe_url%&gt;">Unsub</a> | <a href="&lt;%asm_global_unsubscribe_url%&gt;">ALL</a>
~~~~

So the tags don't get interpreted. This is the recipient's MIME code:

~~~~
<a href=3D"&lt;%asm_preferen=
ces_url%&gt;">Preferences</a> | <a href=3D"&lt;%asm_group_unsubscribe_url%&=
gt;">Unsub</a> | <a href=3D"&lt;%asm_global_unsubscribe_url%&gt;">ALL</a>
~~~~

Based on the discussion above I tried setting the re-setting the tags as a substitution, similar to your sample code, and this is what I get:

~~~~
<a href=3D"<">Preferences</a=
> | <a href=3D"&lt;<&gt;">Unsub</a> | <a href=3D"&lt;<&gt;">ALL</a>
~~~~

I've added the substitution tags as you described. Apologies for the MIME formatting, but as you can see the tag is replaced with a single `<` bracket. So, the substitution is doing something, but not putting in the url. 


In short, I can't save the tags in the template. The suggested work-around of sending them in substitution tags doesn't work.

Thanks for taking the time to look into this. I hope this helps track the cause down!

**Comment by kylearoberts** [comment]
@soundornoise 

Thank you for those details. I know know exactly what you need.

Each tag is designed to to place a link for you in your content. To get the link itself we have different tags:

```
<%asm_preferences_raw_url%>
<%asm_group_unsubscribe_raw_url%>
<%asm_global_unsubscribe_raw_url%>
```

So I can take your example of:

```
  <a href="<%asm_preferences_url%>" >Preferences</a> |
  <a href="<%asm_group_unsubscribe_url%>" >Unsub</a> |
  <a href="<%asm_global_unsubscribe_url%>" >ALL</a>
```

and change it to:

```
  <a href="[Preferences]" >Preferences</a> |
  <a href="[Unsubscribe]" >Unsub</a> |
  <a href="[Unsub_All]" >ALL</a>
```

add in the substitutions:

```
  personalization = Personalization.new
  personalization.to = Email.new(email: 'test1@example.com', name: 'Example User')
  personalization.substitutions = Substitution.new(key: '[Unsubscribe]', value: '<%asm_group_unsubscribe_raw_url%>')
  personalization.substitutions = Substitution.new(key: '[Unsub_All]', value: '<%asm_global_unsubscribe_raw_url%>')
  personalization.substitutions = Substitution.new(key: '[Preferences]', value: '<%asm_preferences_raw_url%>')
  mail.personalizations = personalization
```

This should get you by until we get the encoding issue addresses.

**Comment by soundornoise** [comment]
Yes, I did have the wrong tags. Good spot!

But, I'm still seeing single character substitutions. I get:

~~~~
<a href="%" >Preferences</a> |
<a href="%" >Unsub</a> |
<a href="%" >ALL</a>
~~~~

**Comment by kylearoberts** [comment]
@soundornoise 

Can you provide your template ID and your personalizations? Also, just to make sure are you passing an ASM group in your Ruby?

```
mail.asm = ASM.new(group_id: 99, groups_to_display: [4,5,6,7,8])
```

Once I have this information I should be able to replicate your issue or find what is causing it.

**Comment by rg34** [comment]
@kylearoberts, @soundornoise

Sorry for the delay...just getting back to this. 
The template ID is 8297d3f6-209c-4d1e-ae3d-508d7d34c702,
version 9c7e5728-7726-4e2c-ba6d-458686c68f5a.
And to confirm, I am setting the ASM group in the Ruby code. 

If I update the links in your UI edit template code view with 
<%asm_preferences_url%>
<%asm_global_unsubscribe_url%>,
stay in the code view, save the changes and then send an email from Ruby console, it works. 
We discovered this "workaround" months ago, but it stop working. 

To break it immediately, just save the template in Design view, no other changes necessary, and they will get converted to: 
&lt;%asm_preferences_url%&gt;
&lt;%asm_global_unsubscribe_url%&gt;

Thanks for your help. 
Amy

**Comment by rg34** [comment]
@kylearoberts, @soundornoise 

Just wanted to feed you a quick update. 
After saving the template in code view 12 days ago and since never editing it, I have confirmed that the links are still rendering properly in our emails. Easy to manage on a small team, but definitely hope for a fix in the future.

![screen shot 2017-03-14 at 10 04 17 am](https://cloud.githubusercontent.com/assets/4340339/23909900/9ae8a30e-089d-11e7-9efa-7132a2b2df02.png)
 

Thanks guys.

**Comment by eshanholtz** [comment]
Closing this as it has been marked a non-library issue.
